### ISSUE TITLE
Right to left language text is rendered incorrectly.

### ISSUE THREAD

**Issue by dpeterc** [unixwork/xnedit#131]
Xnedit nicely supports UTF-8 characters, but does not support correct right to left rendering.
Some other text editors like "kate" decide based on first character, and then they render the line left to right or right to left.
See an example of two line text in Persian:
من هر روز از Xnedit استفاده می کنم.
Xnedit بهترین است!

First line is rendered correctly on the web and in kate, second line has Persian text mirrored.
Xnedit has all text mirrored. The text in google translate on the right is rendered correctly in the browser.
![PersianNedit](https://github.com/unixwork/xnedit/assets/15419249/0ab4b42c-5b34-4b5e-b461-9696f36d487e)
First step would be to add right to left rendering on per/line basis, same as in "kate" editor.

Final solution would be to support proper bidi rendering, by using right to left and left to right unprintable UTF-8 markers.
https://en.wikipedia.org/wiki/Right-to-left_mark
https://en.wikipedia.org/wiki/Left-to-right_mark
For text without the markers, some heuristics must still be used, and it will inevitably fail in some cases.

**Comment by unixwork** [comment]
Yes this is completely unimplemented. The NEdit text widget was never designed for that. It will be challenging to implement it but I will look into it.

**Comment by dpeterc** [comment]
I understand that full BIDI is almost impossible, but maybe doing left to right and right to left rendering on per-line basis should be doable. After all, Motif supports it by global setting of XmNlayoutDirection, or to individual widgets or widget classes.
See xnedit with right to left rendering of user interface:
xnedit -xrm "*.layoutDirection: RIGHT_TO_LEFT"

![neditRightToLeft](https://github.com/unixwork/xnedit/assets/15419249/d281bf5f-e765-4fea-9200-e0a3954ffb79)

But I also understand that it is far from the intended use, and that you might wish to concentrate on features useful for programmers, not for translators.

**Comment by Alex-Kent** [comment]
I'm seeing the same behavior even when the entirety of the text is in a RTL script (i.e. RTL scripts are always rendered with LTR character ordering).

[I'm running XNEdit rev 17e59f1 2025-05-08]
